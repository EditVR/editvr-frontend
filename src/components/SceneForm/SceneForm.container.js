/**
 * @file SceneForm.container.js
 * Exports a redux-connected SceneForm component.
 */

import { connect } from 'react-redux';
import { withRouter } from 'react-router-dom';

import SceneForm from './SceneForm';

const mapDispatchToProps = dispatch => ({
  dispatch
});

const mapState = ({ openExperience, user }) => ({
  user,
  experience: openExperience
});

export default connect(mapState, mapDispatchToProps)(withRouter(SceneForm));
