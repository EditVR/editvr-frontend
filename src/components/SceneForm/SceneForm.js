/**
 * @file SceneForm.js
 * Exports a component that allows users to operate on scenes.
 */

import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { TextField, Button, Typography, withStyles } from '@material-ui/core';
import { FileUpload } from '@material-ui/icons';
import { withFormik } from 'formik';
import { string, object } from 'yup';
import Dropzone from 'react-dropzone';

import {
  FORM_BUTTON_INSERT_UPDATE,
  FORM_INPUT_VALIDATION_REGEX_URL_SEGMENT,
  OPEN_EXPERIENCE_SCENE_CREATE,
  OPEN_EXPERIENCE_FETCH_FOR_USER
} from '../../constants';
import SceneFormStyles from './SceneForm.style';
import { Message } from '../';

class SceneForm extends Component {
  static propTypes = {
    classes: PropTypes.shape({
      textField: PropTypes.string.isRequired,
      button: PropTypes.string.isRequired,
      previewImage: PropTypes.string.isRequired
    }).isRequired,
    setFieldValue: PropTypes.func.isRequired,
    handleSubmit: PropTypes.func.isRequired,
    handleChange: PropTypes.func.isRequired,
    isSubmitting: PropTypes.bool.isRequired,
    values: PropTypes.shape({
      title: PropTypes.string,
      body: PropTypes.string,
      field_slug: PropTypes.string
    }).isRequired,
    errors: PropTypes.shape({
      title: PropTypes.oneOfType([PropTypes.string, PropTypes.bool]),
      field_slug: PropTypes.oneOfType([PropTypes.string, PropTypes.bool]),
      body: PropTypes.oneOfType([PropTypes.string, PropTypes.bool])
    }).isRequired,
    experience: PropTypes.shape({
      item: PropTypes.shape({
        field_experience_path: PropTypes.string,
        field_scenes: PropTypes.arrayOf(
          PropTypes.shape({
            id: PropTypes.string.isRequired
          })
        )
      }).isRequired
    }).isRequired,
    user: PropTypes.shape({
      authentication: PropTypes.shape({
        accessToken: PropTypes.string.isRequired,
        csrfToken: PropTypes.string.isRequired
      }).isRequired
    }).isRequired
  };

  state = {
    previewImage: null,
    previewVideo: null
  };

  /**
   * Sets previewImage and previewVideo state.
   */
  previewFile = files => {
    const file = files[0];
    let state = {
      previewImage: file.preview,
      previewVideo: null
    };

    if (file.type.startsWith('video')) {
      state = {
        previewImage: null,
        previewVideo: file.preview
      };
    }

    this.setState(state);
    this.props.setFieldValue('fileName', file.name);

    // Read the file into a binary string and set the sky field.
    const reader = new FileReader();
    reader.readAsDataURL(file);
    reader.onload = () => {
      this.props.setFieldValue('sky', reader.result);
    };
  };

  /**
   * {@inheritdoc}
   */
  render() {
    const {
      classes,
      user: { username },
      experience: { error: apiError, item: experience },
      values,
      errors,
      isSubmitting,
      handleSubmit,
      handleChange
    } = this.props;

    const { previewImage, previewVideo } = this.state;

    // If this form is being rendered when the openExperience state has not
    // yet been loaded, render nothing.
    if (!experience.field_scenes) {
      return null;
    }

    const { field_experience_path: experienceSlug } = experience;

    return (
      <form onSubmit={handleSubmit}>
        {apiError && <Message>{apiError}</Message>}
        <Dropzone onDrop={this.previewFile} className={classes.dropzone}>
          <FileUpload className={classes.dropzoneIcon} />
          <Typography variant="body1" className={classes.dropzoneParagraph}>
            Drag images and drop them here, or click here to upload scene sky
            images (jpg, png) or videos (mp4). These images or videos should be
            equirectangular.
          </Typography>
        </Dropzone>
        {previewImage && (
          <img
            src={previewImage}
            className={classes.previewImage}
            alt="Preview of this scene's sky"
          />
        )}
        {previewVideo && (
          <video controls className={classes.previewImage} src={previewVideo} />
        )}
        {errors.sky && <Message type="error">{errors.sky}</Message>}
        <TextField
          id="title"
          label="Title"
          type="text"
          required
          helperText={
            errors.title
              ? errors.title
              : 'Enter a user-friendly title for your scene.'
          }
          defaultValue={values.title}
          onChange={handleChange}
          error={!!errors.title}
          disabled={isSubmitting}
          className={classes.textField}
        />
        <TextField
          id="field_slug"
          label="URL Path"
          type="text"
          required
          helperText={
            errors.field_slug
              ? errors.field_slug
              : `Enter a url-friendly name for your scene. For example, if you enter 'my-scene', your scene will be published at: /view/${username}/${experienceSlug}/my-scene`
          }
          defaultValue={values.field_slug}
          onChange={handleChange}
          error={!!errors.field_slug}
          disabled={isSubmitting}
          className={classes.textField}
        />
        <TextField
          id="body"
          label="Description"
          type="text"
          multiline
          required
          rows={6}
          helperText={
            errors.body
              ? errors.body
              : 'Describe your experience so people will know what to expect before they enter your scene.'
          }
          defaultValue={values.body}
          onChange={handleChange}
          error={!!errors.body}
          disabled={isSubmitting}
          className={classes.textField}
        />
        <Button
          variant="raised"
          color="primary"
          type="submit"
          className={classes.button}
        >
          {FORM_BUTTON_INSERT_UPDATE}
        </Button>
      </form>
    );
  }
}

const FormikSceneForm = withFormik({
  mapPropsToValues: () => {
    const values = {
      title: '',
      field_slug: '',
      body: '',
      sky: null,
      fileName: null
    };

    return values;
  },
  validationSchema: object().shape({
    title: string()
      .required()
      .min(3)
      .max(50),
    body: string()
      .required()
      .min(3)
      .max(200),
    sky: string()
      .required('You must upload an equirectangular image or video file.')
      .nullable(),
    field_slug: string()
      .required()
      .min(3)
      .max(100)
      .matches(FORM_INPUT_VALIDATION_REGEX_URL_SEGMENT, {
        message:
          'Value must be URL-friendly. No spaces, no special characters, just letters, numbers, and dashes'
      })
  }),
  handleSubmit: (values, { props, setSubmitting }) => {
    const { title, field_slug, body, sky, fileName } = values;
    const {
      dispatch,
      user,
      experience: { item: experience },
      history: { push },
      match: {
        params: { experienceSlug }
      }
    } = props;

    // Extract the file data and construct a payload object.
    dispatch({
      type: OPEN_EXPERIENCE_SCENE_CREATE,
      title,
      body,
      field_slug,
      experience,
      user,
      fileData: sky,
      fileName: `${user.username}-${fileName}`,
      successHandler: () => {
        setSubmitting(false);
        dispatch({
          type: OPEN_EXPERIENCE_FETCH_FOR_USER,
          experienceSlug,
          user
        });

        // Redirect to the newly created scene.
        push(
          `/experience/vreditor/${
            experience.field_experience_path
          }/${field_slug}`
        );
      }
    });
  }
})(SceneForm);

export default withStyles(SceneFormStyles)(FormikSceneForm);
