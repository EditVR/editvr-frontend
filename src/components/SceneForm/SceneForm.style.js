/**
 * @file SceneForm.style.js
 * Exports SceneForm component styles.
 */

export default theme => ({
  textField: {
    width: '100%',
    marginTop: `${theme.spacing.unit * 2}px`
  },
  button: {
    marginTop: `${theme.spacing.unit * 3}px`,
    marginRight: theme.spacing.unit
  },
  dropzone: {
    width: '100%',
    background: theme.palette.primary.light,
    padding: theme.spacing.unit,
    boxSizing: 'border-box',
    marginTop: theme.spacing.unit * 2
  },
  dropzoneParagraph: {
    lineHeight: '1.3',
    display: 'inline-block',
    verticalAlign: 'middle',
    maxWidth: '80%'
  },
  dropzoneIcon: {
    display: 'inline-block',
    verticalAlign: 'middle',
    width: '18%',
    maxWidth: theme.spacing.unit * 10,
    height: 'auto',
    color: theme.palette.common.white
  },
  previewImage: {
    maxWidth: '100%',
    maxHeight: theme.spacing.unit * 20,
    marginTop: theme.spacing.unit * 2
  }
});
