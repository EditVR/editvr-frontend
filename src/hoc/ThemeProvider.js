/**
 * @file ThemeProvider.js
 * Exports a HOC ThemeProvider component that attaches children their MUI theme.
 */

import React from 'react';
import { MuiThemeProvider } from '@material-ui/core';

import muiTheme from '../styles/muiTheme';

const ThemeProvider = WrappedComponent => props => (
  <MuiThemeProvider theme={muiTheme}>
    <WrappedComponent {...props} />
  </MuiThemeProvider>
);

export default ThemeProvider;
