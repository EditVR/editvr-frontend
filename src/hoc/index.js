/**
 * @file index.js
 * Exports all hoc components.
 */

import ThemeProvider from './ThemeProvider/ThemeProvider';
import PrivateRoute from './PrivateRoute/PrivateRoute.container';
import PublicRoute from './PublicRoute/PublicRoute.container';

export { ThemeProvider, PrivateRoute, PublicRoute };
