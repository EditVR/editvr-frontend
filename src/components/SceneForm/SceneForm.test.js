/**
 * @file SceneForm.test.js
 * Contains tests for SceneForm.js.
 */

import React from 'react';
import { mount, shallow } from 'enzyme';
import configureStore from 'redux-mock-store';
import { Provider } from 'react-redux';
import { createMount } from '@material-ui/core/test-utils';
import { MemoryRouter as Router, Route } from 'react-router-dom';

import { ThemeProvider } from '../../hoc';
import SceneForm from './SceneForm.container';

const ThemedSceneForm = ThemeProvider(SceneForm);

describe('<SceneForm />', () => {
  it('Matches its snapshot', () => {
    const store = configureStore()({
      user: {
        error: null,
        authentication: {
          accessToken: 'test',
          csrfToken: 'test'
        }
      },
      openExperience: {
        item: {
          field_experience_path: 'test',
          field_scenes: [
            {
              id: '10'
            }
          ]
        }
      }
    });

    expect(
      createMount()(
        <Provider store={store}>
          <Router initialEntries={['/experience/vreditor/test/test']}>
            <Route
              path="/experience/vreditor/:experienceSlug/:sceneSlug?"
              component={ThemedSceneForm}
            />
          </Router>
        </Provider>
      ).html()
    ).toMatchSnapshot();
  });

  it('Executes its submit handler as expected.', () => {
    const submitFn = jest.fn();
    const store = configureStore()({
      user: {
        error: null,
        authentication: {
          accessToken: 'test',
          csrfToken: 'test'
        }
      },
      openExperience: {
        item: {
          field_experience_path: 'test',
          field_scenes: [
            {
              id: '10'
            }
          ]
        }
      }
    });

    const wrapper = mount(
      shallow(
        shallow(
          <Provider store={store}>
            <Router initialEntries={['/experience/vreditor/test/test']}>
              <Route
                path="/experience/vreditor/:experienceSlug/:sceneSlug?"
                render={() => (
                  <ThemedSceneForm submitHandler={submitFn} store={store} />
                )}
              />
            </Router>
          </Provider>
        ).get(0)
      ).get(0)
    );

    wrapper.find('form').simulate('submit');
    expect(submitFn).toHaveBeenCalledTimes(1);
  });
});
