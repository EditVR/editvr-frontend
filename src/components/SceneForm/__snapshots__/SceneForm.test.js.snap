// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SceneForm /> Matches its snapshot 1`] = `"<form><div class=\\"C-dropzone-3\\" aria-disabled=\\"false\\"><svg class=\\"MuiSvgIcon-root-7 C-dropzoneIcon-5\\" focusable=\\"false\\" viewBox=\\"0 0 24 24\\" aria-hidden=\\"true\\"><g><path d=\\"M9 16h6v-6h4l-7-7-7 7h4zm-4 2h14v2H5z\\"></path></g></svg><p class=\\"MuiTypography-root-13 MuiTypography-body1-22 C-dropzoneParagraph-4\\">Drag images and drop them here, or click here to upload scene sky images (jpg, png) or videos (mp4). These images or videos should be equirectangular.</p><input type=\\"file\\" style=\\"display: none;\\" multiple=\\"\\" autocomplete=\\"off\\"></div><div class=\\"MuiFormControl-root-37 C-textField-1\\" aria-describedby=\\"title-helper-text\\"><label class=\\"MuiFormLabel-root-46 MuiInputLabel-root-41 MuiInputLabel-formControl-42 MuiInputLabel-animated-45\\" data-shrink=\\"false\\" for=\\"title\\">Title<span class=\\"MuiFormLabel-asterisk-50\\"> *</span></label><div class=\\"MuiInput-root-51 MuiInput-formControl-52 MuiInput-underline-55\\"><input aria-invalid=\\"false\\" aria-required=\\"true\\" class=\\"MuiInput-input-59\\" id=\\"title\\" required=\\"\\" type=\\"text\\" value=\\"\\"></div><p class=\\"MuiFormHelperText-root-64\\" id=\\"title-helper-text\\">Enter a user-friendly title for your scene.</p></div><div class=\\"MuiFormControl-root-37 C-textField-1\\" aria-describedby=\\"field_slug-helper-text\\"><label class=\\"MuiFormLabel-root-46 MuiInputLabel-root-41 MuiInputLabel-formControl-42 MuiInputLabel-animated-45\\" data-shrink=\\"false\\" for=\\"field_slug\\">URL Path<span class=\\"MuiFormLabel-asterisk-50\\"> *</span></label><div class=\\"MuiInput-root-51 MuiInput-formControl-52 MuiInput-underline-55\\"><input aria-invalid=\\"false\\" aria-required=\\"true\\" class=\\"MuiInput-input-59\\" id=\\"field_slug\\" required=\\"\\" type=\\"text\\" value=\\"\\"></div><p class=\\"MuiFormHelperText-root-64\\" id=\\"field_slug-helper-text\\">Enter a url-friendly name for your scene. For example, if you enter 'my-scene', your scene will be published at: /view/undefined/test/my-scene</p></div><div class=\\"MuiFormControl-root-37 C-textField-1\\" aria-describedby=\\"body-helper-text\\"><label class=\\"MuiFormLabel-root-46 MuiInputLabel-root-41 MuiInputLabel-formControl-42 MuiInputLabel-animated-45\\" data-shrink=\\"false\\" for=\\"body\\">Description<span class=\\"MuiFormLabel-asterisk-50\\"> *</span></label><div class=\\"MuiInput-root-51 MuiInput-formControl-52 MuiInput-multiline-57 MuiInput-underline-55\\"><textarea aria-invalid=\\"false\\" aria-required=\\"true\\" class=\\"MuiInput-input-59 MuiInput-inputMultiline-61\\" id=\\"body\\" required=\\"\\" rows=\\"6\\" type=\\"text\\"></textarea></div><p class=\\"MuiFormHelperText-root-64\\" id=\\"body-helper-text\\">Describe your experience so people will know what to expect before they enter your scene.</p></div><button tabindex=\\"0\\" class=\\"MuiButtonBase-root-84 MuiButton-root-68 MuiButton-raised-74 MuiButton-raisedPrimary-75 C-button-2\\" type=\\"submit\\"><span class=\\"MuiButton-label-69\\">Save</span><span class=\\"MuiTouchRipple-root-87\\"></span></button></form>"`;
